### DISCUSSION TITLE
Spring Gateway(2.x, 3.x, 4.x) Filter Enhancement in Java agent 9.4(release in 2025Q1)

### DISCUSSION THREAD

**Discussion by wu-sheng** [apache/skywalking#12860] [Show and tell]
_Summarized by [@yqw570994511](https://github.com/yqw570994511)_

This is a new thing started from https://github.com/apache/skywalking/discussions/12839 and PR lands via https://github.com/apache/skywalking-java/pull/736. It enhances the Spring Gateway plugins for GW 2.x, 3.x and 4.x with enhanced context.

Previously, we rely on [Fetch APIs in Webflux Tracing Assistant APIs](https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-webflux/#fetch-trace-context-relative-ids) to obtain relative IDs and print out, but it provides inconsistent format in the logs. The official Trace ID/Context placeholder only prints `N/A`.

With the https://github.com/apache/skywalking-java/pull/736 and released in 9.4, all filters that implement the GlobalFilter and GatewayFilter interfaces can output the logs automatically, meaning without manual codes requirement. And the trace/context IDs are injected into log placeholder.

- Demo codes use `TraceContext.traceId();` and TraceContext.spanId();` for better readings, the log screenshot indicates the consistent outputs.
```java
@Component
public class Filter1 implements GlobalFilter, Ordered {

    private static final Logger log = LoggerFactory.getLogger(Filter1.class);
    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        log.info("filter1 running");

        // fetch trace ID
        String traceId = TraceContext.traceId();
        log.info("filter1 traceId: {}", traceId);

        // fetch segment ID
        String segmentId = TraceContext.segmentId();
        log.info("filter1 segmentId: {}", segmentId);

        // fetch span ID
        int spanId = TraceContext.spanId();
        log.info("filter1 spanId: {}", spanId);

        return chain.filter(exchange);
    }
    @Override
    public int getOrder() {
        return -100;
    }
}
```
```java
@Component
public class GatewayFilter1 implements GatewayFilter {

    private static final Logger log = LoggerFactory.getLogger(GatewayFilter1.class);
    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        log.info("gatewayFilter1 running");
        return chain.filter(exchange);
    }
}
```

The logs and manual APIs output are consistent
![image](https://github.com/user-attachments/assets/c61622c4-2329-42fe-815d-dbb64790c820)

___
But there is still an exceptional case that users still need manual APIs, when they use `chain.filter(exchange).then(...)` as following

- Demo filter codes
```java
@Component
public class Filter2 implements GlobalFilter, Ordered {

    private static final Logger log = LoggerFactory.getLogger(Filter2.class);
    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        log.info("filter2 running");

        // fetch trace ID
        String traceId = TraceContext.traceId();
        log.info("filter2 traceId: {}", traceId);

        // fetch segment ID
        String segmentId = TraceContext.segmentId();
        log.info("filter2 segmentId: {}", segmentId);

        // fetch span ID
        int spanId = TraceContext.spanId();
        log.info("filter2 spanId: {}", spanId);
        return chain.filter(exchange).then(Mono.fromRunnable(() -> {
            String traceId2 = WebFluxSkyWalkingTraceContext.traceId(exchange);
            log.info("filter2 then traceId: {}", traceId2);

            String segmentId2 = WebFluxSkyWalkingTraceContext.segmentId(exchange);
            log.info("filter2 then segmentId: {}", segmentId2);

            int spanId2 = WebFluxSkyWalkingTraceContext.spanId(exchange);
            log.info("filter2 then spanId: {}", spanId2);
        }));
    }

    @Override
    public int getOrder() {
        return 10;
    }
}

```
- Output logs without proper trace context, but manual APIs work well.
<img width="1364" alt="image" src="https://github.com/user-attachments/assets/587edde5-f774-40ff-9b94-c8387646e83f" />

**Comment by hyhyf** [comment]
Need more testing?
My project is using java agent 9.0.0, which can print the correct trace id.

**Reply by hyhyf** [reply]
> Need more testing? My project is using java agent 9.0.0, which can print the correct trace id.

Using the apm-spring-cloud-gateway-3.x-plugin

**Reply by wu-sheng** [reply]
We can't tell from the old version. What we do is discussing for the latest
